extensions/wikia/AdEngine/js/provider/gpt/GoogleTag.js

```diff
@@ -0,0 +1,131 @@
+/*global define,setTimeout*/
+/*jshint maxlen:125, camelcase:false, maxdepth:7*/
+define('ext.wikia.adEngine.provider.googleTag', [
+	'wikia.document',
+	'wikia.log',
+	'wikia.window'
+], function (doc, log, window) {
+
+	var logGroup = 'ext.wikia.adEngine.provider.googleTag',
+		initialized = false,
+		registeredCallbacks = {},
+		slots = {},
+		slotQueue = [],
+		googleTag,
+		pubAds;
+
+	function init() {
+		log('init', 'debug', logGroup);
+
+		var gads = doc.createElement('script'),
+			node = doc.getElementsByTagName('script')[0];
+
+		window.googletag = window.googletag || {};
+		window.googletag.cmd = window.googletag.cmd || [];
+
+		gads.async = true;
+		gads.type = 'text/javascript';
+		gads.src = '//www.googletagservices.com/tag/js/gpt.js';
+
+		log('Appending GPT script to head', 'debug', logGroup);
+
+		node.parentNode.insertBefore(gads, node);
+		googleTag = window.googletag;
+
+		// Enable services
+		log(['init', 'googletag.cmd.push'], 'info', logGroup);
+		googleTag.cmd.push(function () {
+			pubAds = googleTag.pubads();
+
+			pubAds.collapseEmptyDivs();
+			pubAds.enableSingleRequest();
+			pubAds.disableInitialLoad(); // manually request ads using refresh
+			pubAds.addEventListener('slotRenderEnded', dispatchEvent);
+
+			googleTag.enableServices();
+
+			log(['init', 'googletag.cmd.push', 'done'], 'debug', logGroup);
+		});
+		initialized = true;
+	}
+
+	function isInitialized() {
+		return initialized;
+	}
+
+	function push(callback) {
+		googleTag.cmd.push(callback);
+	}
+
+	function flush() {
+		if (!isInitialized()) {
+			log(['flush', 'done', 'No slots to flush'], 'info', logGroup);
+			return;
+		}
+
+		googleTag.cmd.push(function () {
+			log(['flush', 'start'], 'info', logGroup);
+
+			log(['flush', 'refresh', slotQueue], 'debug', logGroup);
+			if (slotQueue.length) {
+				pubAds.refresh(slotQueue);
+				slotQueue = [];
+			}
+
+			log(['flush', 'done'], 'info', logGroup);
+		});
+	}
+
+	function addSlot(slotName, slotPath, slotTargeting, element) {
```

Comment:
addSlot should only accept an instance of AdElement as the param.
AdElement should know about slotPath and slotTargeting.